### DISCUSSION TITLE
C# stack and async functions

### DISCUSSION THREAD

**Discussion by Dunge** [pulumi/pulumi#9495] [Pulumi in Practice] [answered]
I'm trying to follow the sample [here ](https://www.pulumi.com/docs/guides/crosswalk/aws/eks/) that end up calling `Ecr.GetCredentials.Invoke` to get Ecr credentials.

I see that the new method is now `InvokeAsync`. I also see a mention [here](https://www.pulumi.com/docs/get-started/install/migrating-2.0/) saying that we only need to call this async version instead and await it.

My problem is that the whole Stack definition code is done inside a class constructor, which can't be set as async in C#.

The program.cs file seems ready to support it:
`static Task<int> Main() => Deployment.RunAsync<MyStack>();`

I'm just wondering how to use it to have `MyStack` capable of awaiting stuff in his stack definition.

**Comment by karpikpl** [comment, marked as answer]
I think that's a valid question. 
New templates don't use stacks but async function instead.

I've ended up doing this:
```csharp
return await Pulumi.Deployment.RunAsync(() => new CoreEx.Infra.CoreExStack().ExecuteStackAsync(), null);
```
where CoreExStack no longer inherits from `Stack`

**Reply by Dunge** [reply]
Thanks. I decided not to use Pulumi for my project after all, but I'm glad to at least see an answer.
